node/src/store.rs

```diff
@@ -1,9 +1,136 @@
+            height_to_hash: DashMap::new(),
+        }
+    }
+
+    pub fn get_head_height(&self) -> usize {
+        self.height_to_hash.len()
+    }
+
+    pub fn append_header(&self, header: ExtendedHeader) -> Result<(), InsertError> {
+        let hash = header.hash();
+        let height = header.height(); // TODO: should Store be responsible for making sure we have
+                                      // all 1..=head headers?
+
+        // lock both maps to ensure consistency
+        // this shouldn't deadlock as long as we don't hold references across awaits if any
+        // https://github.com/xacrimon/dashmap/issues/233
+        let hash_entry = self.headers.entry(hash);
+        let height_entry = self.height_to_hash.entry(height.into());
+
+        if matches!(hash_entry, Entry::Occupied(_)) {
+            return Err(InsertError::HashExists(hash));
+        }
+        if matches!(height_entry, Entry::Occupied(_)) {
+            return Err(InsertError::HeightExists(height.into()));
+        }
+
+        info!("Will insert {hash} at {height}");
+        hash_entry.insert(header);
+        height_entry.insert(hash);
+
+        Ok(())
+    }
+
+    pub fn get_head(&self) -> Result<ExtendedHeader, ReadError> {
+        let head_height = match self.height_to_hash.len() {
+            0 => return Err(ReadError::NotFound),
+            height => match u64::try_from(height) {
+                Ok(h) => h,
+                Err(e) => {
+                    error!("Failed to convert {height} from usize to u64, should not happen: {e}");
+                    return Err(ReadError::HeadHeightConversionError);
+                }
+            },
+        };
+
+        let Some(head_hash) = self.height_to_hash.get(&head_height).map(|v| *v) else {
+            error!("height_to_hash[height_to_hash.len()] not found, store is inconsistent");
+            return Err(ReadError::LostStoreHead);
+        };
+
+        match self.headers.get(&head_hash) {
+            Some(v) => Ok(v.clone()),
+            None => {
+                error!("Header with hash {head_hash} for height {head_height} missing");
+                Err(ReadError::LostHash)
+            }
+        }
+    }
+
+    pub fn get_by_hash(&self, hash: &Hash) -> Result<ExtendedHeader, ReadError> {
+        match self.headers.get(hash) {
+            Some(v) => Ok(v.clone()),
+            None => Err(ReadError::NotFound),
+        }
+    }
+
+    pub fn get_by_height(&self, height: u64) -> Result<ExtendedHeader, ReadError> {
+        let Some(hash) = self.height_to_hash.get(&height).map(|h| *h) else {
+            return Err(ReadError::NotFound);
+        };
+
+        match self.headers.get(&hash) {
+            Some(h) => Ok(h.clone()),
+            None => {
+                error!("Lost hash {hash} at height {height}");
+                Err(ReadError::LostHash)
+            }
+        }
+    }
+
+    #[doc(hidden)]
+    pub fn test_filled_store(height: u64) -> Self {
```

Comment:
This should be in `mod tests` instead.

Reply:
Created `test_utils` and moved this and `gen_extended_header` there.